### PR TITLE
Remove wildcard defaults for app, db, node, and kubernetes labels

### PR DESCRIPTION
**Pull request by nklaassen**
This change removes the default wildcard permissions for `app_labels`, `db_labels`, `node_labels`, and `kubernetes_labels` currently set in `CheckAndSetDefaults` for all roles where those labels are not specified. As mentioned in #5513, this can result in accidentally granting access to all of these resources for roles which are not meant to.

To allow the built-in roles access to these resources, I modified `GetCheckerForBuiltinRole` to grant the wildcard permissions which were needed to pass the existing tests, as many of the built-in roles currently rely on the permissive defaults.

This change will not affect existing roles which have already been persisted with the wildcard values, only newly created roles.

### CODE REVIEW

File: lib/auth/permissions.go
Comment (klizhentas) [inline lib/auth/permissions.go:529]: Are we sure that all those builtins really need a wildcard? What would break if we not add those lines?

**Comment by klizhentas** [comment]
@nklaassen welcome to the team 👋 and congrats on the first PR 🚀

File: lib/auth/permissions.go
Comment (nklaassen) [reply lib/auth/permissions.go:529]: Many of the tests break without these, I only added the labels which were necessary to fix the tests. I'm not really adding any permissions here, just making more explicit what we are already relying on from the current defaults.

I don't have a deep understanding of these built-in roles yet, so it could just a problem with the tests using these built-in roles incorrectly rather than these permissions being fundamentally necessary. But it seems logical that the db role would need database access, the proxy role would need access to everything, etc. If I'm missing something, we may need to discuss this further.

File: api/types/role.go
Comment (andrejtokarcik) [inline api/types/role.go:591]: Remove these `if`s entirely...?

Also, shouldn't the `NodeLabels` wildcard above these be got rid of as well?

**Review by andrejtokarcik — COMMENTED** [review]
What about the necessary migrations we discussed in https://github.com/gravitational/teleport/issues/5513? Is this change going to be included in a release where we can already assume all the roles are explicitly granted the labels they need?

**Comment by andrejtokarcik** [comment]
Oh, ignore my migration question above. The old roles should already be stored with the wildcard patterns so it's not a problem, right?

Please also mention `db_labels` in the PR/commit description.

**Comment by nklaassen** [comment]
> Oh, ignore my migration question above. The old roles should already be stored with the wildcard patterns so it's not a problem, right?

That's right.

> Please also mention `db_labels` in the PR/commit description.

Done, as mentioned above I had accidentally written `node_labels` instead of `db_labels` in the description. `node_labels` are currently unchanged in this PR.

File: lib/auth/permissions.go
Comment (r0mant) [inline lib/auth/permissions.go:301]: Hmm, I don't think RoleNode (which represents an SSH node) should have AppLabels set?

File: api/types/role.go
Comment (r0mant) [inline api/types/role.go:581]: What about NodeLabels? Do we also want to default to nothing there as well, or would it break any use-cases?

File: lib/auth/permissions.go
Comment (r0mant) [inline lib/auth/permissions.go:380]: What about KubernetesLabels?

File: lib/auth/permissions.go
Comment (r0mant) [inline lib/auth/permissions.go:440]: KubernetesLabels?

File: lib/auth/permissions.go
Comment (r0mant) [inline lib/auth/permissions.go:534]: KubernetesLabels?

File: lib/auth/permissions.go
Comment (nklaassen) [reply lib/auth/permissions.go]: Good point, I had thought there was a test that wanted this but apparently not, I've removed it now.

File: api/types/role.go
Comment (nklaassen) [reply api/types/role.go:581]: I've removed the wildcards for NodeLabels as well in the latest rev. It doesn't break any of the tests, and I don't think it would break any usecases.

File: lib/auth/permissions.go
Comment (nklaassen) [reply lib/auth/permissions.go]: Originally I was only adding the permissions that were required to make the tests pass, but this would make sense I think with the new defaults. Done.

File: api/types/role.go
Comment (nklaassen) [reply api/types/role.go]: I've now removed the node_labels wildcards as well.

**Comment by klizhentas** [comment]
@russjones @fspmarshall @awly because of potential importance of this PR, asking for a couple of additional reviews on this

**Review by fspmarshall — COMMENTED** [review]
This is still a breaking change since it will break automation that creates teleport roles. IMO we will need to bump the role version number and continue to apply the current defaults when dealing with `v2` roles.

Unfortunately, changes to roles that appear as errors to older teleport instances (unknown version number in this case) can effectively "brick" those instances under some circumstances (or at least, used to be able to... I haven't tried this recently).  At the very least, unexpected resource versions cause infinite re-init loops in the cache, which consume a bunch of network resources (unlike adding a new resource kind, or a new field to an existing resource, which are changes that can be ignored by older versions).

Given the above, I'd suggest that we either have an RFD process for a `v3` role, and try to get as many future-looking improvements packed in there as possible, or come up with a "minor version bump" scheme for role resources (e.g. `v2.1`) which provides some kind of compatibility guarantees so that older teleport versions can know if they can reasonably consume the resources (e.g. in this case, an older node could safely consume this resource so long as the auth server had ensured that the labels were empty mappings as opposed to being `nil`, since that wouldn't trigger the node to insert wildcards).

**Comment by nklaassen** [comment]
@fspmarshall good point about this potentially breaking automated role creation. I agree we may need to put more thought into this, including bumping the role version and everything that comes with that. I will bring this up with @russjones in our sync today.

File: api/types/role.go
Comment (nklaassen) [reply api/types/role.go]: Good point, I've gone back to setting the empty label set in the latest commit.

File: lib/auth/permissions.go
Comment (russjones) [inline lib/auth/permissions.go:325]: Can you remove this line and share the test that fails? I wonder if we're dealing with an incorrect test or incorrect implementation.

Comment (nklaassen) [reply lib/auth/permissions.go:325]: @russjones With this removed, it's TestAuthTokens and TLSSuite.TestAppServerCRUD that are failing

File: lib/auth/permissions.go
Comment (nklaassen) [inline lib/auth/permissions.go:237]: This fixes TestAppAccessForward, does this permission seem reasonable?

**Comment by russjones** [comment]
@fspmarshall Valid points. As you mentioned moving to `RoleV4` is not simple. We have to continue returning `RoleV3` for a release (Teleport 7) for older clients and then fully move off in the release after that (Teleport 8) at which point users will still face a breaking change during unmarshal as we drop support for `RoleV3`. It's more a matter of when the breaking change will occur not if.

However, since we're not actually changing the schema of the role and changing the behavior, this feels more like a bug fix to me. In that situation, I think it's reasonable to include this change in Teleport 7 with a notice about changed behavior.

@klizhentas what do you think?

**Comment by klizhentas** [comment]
@russjones @nklaassen sorry for the delay, I will review tomorrow

File: lib/auth/permissions.go
Comment (klizhentas) [reply lib/auth/permissions.go:249]: Unclear how this would have any impact, because the `Rules` section does not list KindApp

File: lib/auth/permissions.go
Comment (klizhentas) [reply lib/auth/permissions.go:338]: Unclear why App should be able to read all apps. Probably App should be able to read it's own record instead.

File: lib/auth/permissions.go
Comment (klizhentas) [inline lib/auth/permissions.go:364]: Same here. DB should be able to read their own record instead of this clause. See auth_with_roles `GetUser` for an example of an exception clause.

File: lib/auth/permissions.go
Comment (klizhentas) [inline lib/auth/permissions.go:572]: KubeService should not be able to write all other kube service records, that would be a security breach.

**Review by klizhentas — CHANGES_REQUESTED** [review]
@fspmarshall @russjones I'm ok to break the behavior of existing role labels with wildcards, as this is a bug we have to fix and the features are relatively new. However, there are some security problems with this PR we have to address first.

**Comment by fspmarshall** [comment]
> I'm ok to break the behavior of existing role labels with wildcards, as this is a bug we have to fix and the features are relatively new.

@sasha As currently implemented, this PR also changes the default for node labels, which has been defaulting to wildcard allow for at least 4 years.  Are you saying change the default for everything *but* node labels?

File: lib/auth/permissions.go
Comment (nklaassen) [reply lib/auth/permissions.go:364]: Makes sense, I can work on that.

File: lib/auth/permissions.go
Comment (nklaassen) [reply lib/auth/permissions.go:572]: That also makes sense. Just to be clear though, if that would be a security breach then it already exists today - all of these roles currently have the wildcard labels applied in `CheckAndSetDefaults`

**Review by awly — COMMENTED** [review]
I agree with @klizhentas, let's move all the logic for builtin roles (services) into `lib/auth/auth_with_roles.go`. Label checks should probably not apply to builtin services, only to human users.

Today, any service can mess with any other service of the same kind (e.g. kube service can read or create any other kube service). This is more permissive than necessary.
The logic can be very similar to https://github.com/gravitational/teleport/blob/991059869efe1019d79ff50a5885c53baea9571c/lib/auth/auth_with_roles.go#L81-L90, and only apply to builtin roles:
- write: any builtin service can only write their own kind and name (e.g. kube service can only write a kube service of the same name as specified in its client cert)
- read:
  - auth and proxy can read anything
  - kube/app/node/db can only read their own instance

To keep this PR from scope creep, let's do this:
- split authz checks in `auth_with_roles.go` between users and builtin roles when accessing labeled resources
- for users, check RBAC verbs and labels as usual
- for builtin roles, only check RBAC verbs and skip label check - this preserves the current behavior
- after this merges, create a separate PR that further restricts builtin roles as described above

Does that sound reasonable?

**Comment by klizhentas** [comment]
agree with @awly's comment 💯

File: lib/services/role_test.go
Comment (nklaassen) [reply lib/services/role_test.go]: Good point, it probably makes more sense to set the defaults on all of the roles in this test actually

**Comment by nklaassen** [comment]
@awly that sounds reasonable, I'll reach out to you for more details

**Comment by nklaassen** [comment]
Okay I'm working on @awly's suggested change to avoid having the wildcards in the built-in roles.

I don't want to lose the issue of `node_labels` though - @andrejtokarcik had suggested maybe removing the current wildcard default for users with logins (which I've done in this PR now), while it seems like @fspmarshall is thinking it should stay as it's been there for 4 years. @klizhentas @russjones @awly what do you think?

File: lib/auth/permissions.go
Comment (nklaassen) [reply lib/auth/permissions.go:249]: > Unclear how this would have any impact, because the `Rules` section does not list KindApp

The default implicit role has a rule for RO access to KindAppServer, which allows this RoleSet to pass the actions check but potentially fail the labels check

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: lib/auth/permissions.go
Comment (awly) [reply lib/auth/permissions.go:726]: @klizhentas I suspect that a lot of basic functionality could break if a proxy can't read the list of ssh/k8s/db/webapp instances

The idea was to tighten this part with role-based exception cases in a followup PR. No actual release would include this blanked exception as is.

**Review by awly — APPROVED** [review]

File: lib/auth/permissions.go
Comment (awly) [reply lib/auth/permissions.go:726]: * connect access - fails because builtin roles don't have `logins`, `kube_users`, `db_users`, etc
* write access - subject to RBAC write verbs as before; labels didn't apply there in the first place
* read/list access - subject to RBAC read verbs as before; but you can now read SSH nodes as any service (not just proxy, due to custom logic in `filterNodes`)

**Comment by r0mant** [comment]
@awly @russjones @klizhentas @nklaassen We have a customer who's waiting for this fix and is blocked from adopting application access until this is addressed. Currently, since this is a change in behavior, it is scheduled for 7.0 which is about 2 months away (at least). Any way we can backport this (once it's done) to branch/v6 sooner, so it makes it into 6.3 for example? Maybe not fully, but at least parts related to application/database access since they're newer and it is less likely someone depends on the wildcard behavior (at least intentionally) for those. Thanks.

**Review by fspmarshall — CHANGES_REQUESTED** [review]
> Any way we can backport this (once it's done) to branch/v6 sooner, so it makes it into 6.3 for example?

IMO this cannot be done without putting the new defaults behind some kind of gate.  Breaking existing configuration and documentation in the middle of a major release cycle isn't acceptable.


I am also *strongly* of the opinion that we cannot change the default behavior of `node_labels` for `V3` roles, even with a major release.  This would break automation and documentation that has worked for more than four years.  Not just our documentation, but any internal documentation that our users may have developed over the years.

I think that both the problem of releasing the elements this behavior to those who want it before `7.0`, and the problem of maintaining backwards compatibility with legacy automation and documentation, can be solved by addding support for minor versions to roles.  I.e. support `v3.1` in the `version` field.  In order to support backwards compatibility with older nodes, we could add a temporary shim which caused the event system and role getters to map `v3.1` roles to `v3` roles once their defaults have been set.

**Comment by awly** [comment]
@r0mant this is a pretty significant breaking change so I'm also against doing any backports to 6.3.

@fspmarshall I think we need @klizhentas to make the call about `node_labels` and/or any `Role` API versioning.

**Comment by r0mant** [comment]
@awly @fspmarshall @klizhentas @russjones

If we go with `v3.1` approach, should be safe to backport?

Also, what if we did something like this: introduce `v3.1` role version which would only apply on create i.e.:
* Default labels won't be set to wildcards.
* But when inserting into the database, convert version to `v3`.

This way, cache won't be affected since role version will be `v3` but users who wish to not use default wildcard labels will define their roles as `v3.1`. Any reason this wouldn't work?

**Comment by russjones** [comment]
This PR has been good to explore the issues surrounding changes that need to be made and the potential issues, but I think we need a quick sync meeting and RFD to continue this.

So I am going to close it for now and will follow up with relevant parties a bit later.

**Pull request closed without merging**
